tests/std/tests/GH_003131_tuple_opeq_overload_and/test.cpp

```diff
@@ -0,0 +1,33 @@
+#include <cassert>
+#include <tuple>
+
+using namespace std;
+
+struct EvilBool {
+    bool b;
+
+    operator bool() const {
+        return b;
+    }
+    EvilBool operator!() const;
+
+    // EVILLLLL
+    friend bool operator&&(const EvilBool lhs, bool rhs);
+    friend bool operator||(const EvilBool lhs, bool rhs);

```

Comment:
I believe (with varying degrees of confidence) that (1) this test case is technically non-conforming, (2) there is a potential LWG issue to file here, (3) the product code should probably not be changed.

1. This test case is technically non-conforming - I believe this is the clearest case. Lines 31-32 use `op<` and `op>` for `tuple`, which is spaceship-powered in C++20. WG21-N4917 [tuple.rel]/6 says that `tuple` spaceship uses *`synth-three-way`*. [expos.only.func]/2 says that *`synth-three-way`* requires that `t < u` returns *`boolean-testable`*. [concept.booleantestable]/1 explains "The exposition-only *`boolean-testable`* concept specifies the requirements on expressions that are convertible to `bool` and for which the logical operators (7.6.14, 7.6.15, 7.6.2.2) have the conventional semantics.", /2.2 and related Standardese forbids evil `operator&&` and `operator||`, and /6 summarizes: "[*Note 1:* The intention is to ensure that given two types `T1` and `T2` that each model *`boolean-testable-impl`*, the `&&` and `||` operators within the expressions `declval<T1>() && declval<T2>()` and `declval<T1>() || declval<T2>()` resolve to the corresponding built-in operators. *— end note*]"

2. There is a potential LWG issue to file here. I believe that `tuple` `operator<=>` clearly invokes the *`boolean-testable`* prohibition of evil logical operators. The `tuple` `operator==` specification (which is TR1-era and has been massively updated over the years) doesn't say *`boolean-testable`*, [tuple.rel]/2 just says "`get<i>(t) == get<i>(u)` is a valid expression returning a type that is convertible to `bool`". But before concluding that it should tolerate evil logical operators, look at `pair`: [pairs.spec]/1 "*Returns:* `x.first == y.first && x.second == y.second`." (As a general rule, `tuple` and `pair` try to follow the same behavior, with only a few differences like how `pair` comparisons traditionally required identical types, but `tuple` permits variation - arguably `pair` should be upgraded there.) I think that the Standardese is clearly inconsistent here, although not quite to the point of self-contradiction (`tuple` `<` ends up requiring that element `<` returns *`boolean-testable`*, but `tuple` `==` doesn't require element `==` to be *`boolean-testable`*, while `pair` `==` effectively requires `&&` to behave reasonably). I believe that this merits an LWG issue, and should be resolved by making `tuple` `operator==` require *`boolean-testable`*.

3. The product code should not be changed here - I believe that given the Standardese for `tuple` `operator<` (where evil logical operators are forbidden), and the potential LWG issue in `tuple` `operator==`, that we shouldn't be attempting to defend against evil logical operators here.

Reply:
> This test case is technically non-conforming

> I think that the Standardese is clearly inconsistent here

I believe both are true only since C++20. Although there's potential DR WG21-P2167 which resolves LWG-2114 and possibly amends `operator==` for `tuple` and `pair`, but it's not completed as for now.

1. The requirements of _`boolean-testable`_ are introduced here alongwith `operator<=>`. So the test case should be considered conforming in C++14/17, where the return types are only required to be implicitly convertible to `bool`.
2. These operators used to be consistently required before C++20, see WG21-N4659 [[tuple.rel]/1](https://timsong-cpp.github.io/cppwp/n4659/tuple.rel#1), [[tuple.rel]/4](https://timsong-cpp.github.io/cppwp/n4659/tuple.rel#4).


> while `pair` `==` effectively requires `&&` to behave reasonably

It's a bit strange that the `operator==` (and `operator<` until C++20, see WG21-N4659 [[pairs.spec]/2](https://timsong-cpp.github.io/cppwp/n4659/pairs.spec#2)) doesn't have any additional requirements, which means that `operator==` for elements might have weird return types, and shortcut evaluation is not guaranteed.


> The product code should not be changed here

I guess the product code was written before the completion of C++14 `constexpr` mechanisms (WG21-N3652). So it seems arguable to me whether the product code should be change.

I prefer [Optional B for `pair`](https://wg21.link/p2167r2#pairs.spec_opt_b) and [Optional A for `tuple`](https://wg21.link/p2167r2#tuple.rel_opt_a) in P2167R2.
It may be more reasonable to mandate (the diagnostic-requiring portion of) _`boolean-testable`_ for `pair` and `tuple`, but these options might require more changes, and it's difficult to apply them to C++14/17 modes.

Reply:
Please, let's have any further discussion about whether or not this issue is a bug that we need to address in #3131 so we don't have separate discussions.